File: Java/linkedListCycle_141.java

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class linkedListCycle_141 {
    public boolean hasCycle(ListNode head) {
        if(head == null || head.next == null) {
            return false;
        }
        ListNode first = head;
        ListNode second = head;
        while(second.next != null && second.next.next != null) {
            first = first.next;
            second = second.next.next;
            if(first == second) {
                return true;
            }
        }
        return false;
    }
    // Time Complexity: O(n)
    // Space Complexity: O(1)
}

File: Cpp/linked_list_cycle_141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr) {
            return false;
        }
        ListNode* first = head;
        ListNode* second = head;
        while(first->next != nullptr && second->next != nullptr && second->next->next != nullptr) {
            first = first->next;
            second = second->next->next;
            if(first == second) {
                return true;
            }
        }
        return false;
    }
};

// Solution 2
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* first = head;
        ListNode* second = head;
        while(second != nullptr && second->next != nullptr) {
            first = first->next;
            second = second->next->next;
            if(first == second) {
                return true;
            }
        }
        return false;
    }
};